File: generator.py
from opcodes import OpCode, Instruction
from parser import (
    ASTNode,
    Literal,
    Identifier,
    ArrayLiteral,
    ArrayAccess,
    FunctionCall,
    BinaryOp,
    UnaryOp,
    Assignment,
    IfStatement,
    WhileLoop,
    ForLoop,
    RepeatUntilLoop,
    PrintStatement,
    ReturnStatement,
    Block,
    FunctionStatement,
)


class Generator:
    """Generates machine-level opcodes from AST"""

    def __init__(self):
        self.instructions = []
        self.temp_counter = 0
        self.label_counter = 0
        self.function_table = {}  # name -> (start_addr, [param_names])
        self.entry_params = []  # params of the first / entry function

    def new_temp(self) -> str:
        temp = f"t{self.temp_counter}"
        self.temp_counter += 1
        return temp

    def new_label(self) -> str:
        label = f"L{self.label_counter}"
        self.label_counter += 1
        return label

    def emit(self, opcode: OpCode, *operands):
        self.instructions.append(Instruction(opcode, *operands))

    def generate(self, ast: Block) -> list[Instruction]:
        self.instructions = []
        self.temp_counter = 0
        self.label_counter = 0
        self.function_table = {}
        self.entry_params = []

        functions = []
        bare_stmts = []
        for stmt in ast.statements:
            if isinstance(stmt, FunctionStatement):
                functions.append(stmt)
            else:
                bare_stmts.append(stmt)

        if functions:
            # First function is the entry point; compile its body first
            entry = functions[0]
            self.entry_params = [p.name for p in entry.param_ids]
            entry_start = len(self.instructions)
            self.function_table[entry.name.name] = (entry_start, self.entry_params)
            self.visit(entry.body)
            self.emit(OpCode.RET)

            # Helper functions follow in the instruction stream
            for func in functions[1:]:
                func_start = len(self.instructions)
                params = [p.name for p in func.param_ids]
                self.function_table[func.name.name] = (func_start, params)
                self.visit(func.body)
                self.emit(OpCode.RET)

        # Bare top-level statements (no Algorithm wrapper)
        for stmt in bare_stmts:
            self.visit(stmt)

        return self.instructions

    def visit_block(self, node: Block):
        for stmt in node.statements:
            self.visit(stmt)

    def visit(self, node: ASTNode) -> str | None:
        if isinstance(node, Literal):
            return self.visit_literal(node)
        elif isinstance(node, Identifier):
            return self.visit_identifier(node)
        elif isinstance(node, ArrayAccess):
            return self.visit_array_access(node)
        elif isinstance(node, FunctionCall):
            return self.visit_function_call(node)
        elif isinstance(node, BinaryOp):
            return self.visit_binary_op(node)
        elif isinstance(node, UnaryOp):
            return self.visit_unary_op(node)
        elif isinstance(node, Assignment):
            return self.visit_assignment(node)
        elif isinstance(node, IfStatement):
            return self.visit_if_statement(node)
        elif isinstance(node, WhileLoop):
            return self.visit_while_loop(node)
        elif isinstance(node, ForLoop):
            return self.visit_for_loop(node)
        elif isinstance(node, RepeatUntilLoop):
            return self.visit_repeat_until_loop(node)
        elif isinstance(node, PrintStatement):
            return self.visit_print_statement(node)
        elif isinstance(node, ReturnStatement):
            return self.visit_return_statement(node)
        elif isinstance(node, FunctionStatement):
            return self.visit_function_statement(node)
        elif isinstance(node, Block):
            self.visit_block(node)
            return None
        else:
            raise ValueError(f"Unknown AST node type: {type(node)}")

    def visit_literal(self, node: Literal) -> str:
        return str(node.value)

    def visit_identifier(self, node: Identifier) -> str:
        return node.name

    def visit_array_access(self, node: ArrayAccess) -> str:
        array_name = (
            node.array.name
            if isinstance(node.array, Identifier)
            else self.visit(node.array)
        )
        index = self.visit(node.index)
        temp = self.new_temp()
        self.emit(OpCode.IDX, array_name, index, temp)
        return temp

    def visit_function_call(self, node: FunctionCall) -> str:
        arg_vals = [self.visit(arg) for arg in node.args]
        temp = self.new_temp()
        self.emit(OpCode.CAL, node.name, *arg_vals, temp)
        return temp

    def visit_binary_op(self, node: BinaryOp) -> str:
        left = self.visit(node.left)
        right = self.visit(node.right)
        temp = self.new_temp()

        if node.operator in ["+", "-", "*", "/", "mod"]:
            self.emit(OpCode.AOP, node.operator, left, right, temp)
        elif node.operator in ["=", "<", ">", "<=", ">=", "!="]:
            self.emit(OpCode.COM, node.operator, left, right, temp)
        elif node.operator in ["and", "or"]:
            self.emit(OpCode.COM, node.operator, left, right, temp)
        else:
            raise ValueError(f"Unknown operator: {node.operator}")

        return temp

    def visit_unary_op(self, node: UnaryOp) -> str:
        operand = self.visit(node.operand)
        temp = self.new_temp()
        # 3-operand AOP signals unary operation to the VM
        self.emit(OpCode.AOP, node.operator, operand, temp)
        return temp

    def visit_assignment(self, node: Assignment) -> None:
        value = self.visit(node.value)

        if isinstance(node.target, Identifier):
            self.emit(OpCode.ASN, node.target.name, value)
        elif isinstance(node.target, ArrayAccess):
            array_name = (
                node.target.array.name
                if isinstance(node.target.array, Identifier)
                else self.visit(node.target.array)
            )
            index = self.visit(node.target.index)
            self.emit(OpCode.ASN, f"{array_name}[{index}]", value)
        else:
            raise ValueError(f"Invalid assignment target: {type(node.target)}")

    def visit_if_statement(self, node: IfStatement) -> None:
        self.visit(node.condition)

        skp_idx = len(self.instructions)
        self.emit(OpCode.SKP, 0)

        then_start = len(self.instructions)
        self.visit(node.then_block)

        if node.else_block:
            jmp_idx = len(self.instructions)
            self.emit(OpCode.JMP, 0)

            self.visit(node.else_block)

            then_size = jmp_idx - then_start + 1
            self.instructions[skp_idx] = Instruction(OpCode.SKP, then_size)

            end_idx = len(self.instructions)
            self.instructions[jmp_idx] = Instruction(OpCode.JMP, end_idx)
        else:
            then_size = len(self.instructions) - then_start
            self.instructions[skp_idx] = Instruction(OpCode.SKP, then_size)

    def visit_while_loop(self, node: WhileLoop) -> None:
        loop_start = len(self.instructions)

        self.visit(node.condition)

        skp_idx = len(self.instructions)
        self.emit(OpCode.SKP, 0)

        body_start = len(self.instructions)
        self.visit(node.body)

        self.emit(OpCode.JMP, loop_start)

        body_size = len(self.instructions) - body_start
        self.instructions[skp_idx] = Instruction(OpCode.SKP, body_size)

    def visit_for_loop(self, node: ForLoop) -> None:
        """for VAR <- INIT (to|downto) END [by STEP] do BODY end"""
        self.visit(node.assignment)

        loop_var = (
            node.assignment.target.name
            if isinstance(node.assignment.target, Identifier)
            else str(node.assignment.target)
        )

        loop_start = len(self.instructions)
        end_value = self.visit(node.end)

        cmp_op = "<=" if node.direction == "to" else ">="
        temp_cmp = self.new_temp()
        self.emit(OpCode.COM, cmp_op, loop_var, end_value, temp_cmp)

        skp_idx = len(self.instructions)
        self.emit(OpCode.SKP, 0)

        body_start = len(self.instructions)
        self.visit(node.body)

        # Increment / decrement
        temp_inc = self.new_temp()
        if node.step is not None:
            step_val = self.visit(node.step)
            self.emit(OpCode.AOP, "+", loop_var, step_val, temp_inc)
        elif node.direction == "to":
            self.emit(OpCode.AOP, "+", loop_var, "1", temp_inc)
        else:
            self.emit(OpCode.AOP, "-", loop_var, "1", temp_inc)

        self.emit(OpCode.ASN, loop_var, temp_inc)
        self.emit(OpCode.JMP, loop_start)

        body_size = len(self.instructions) - body_start
        self.instructions[skp_idx] = Instruction(OpCode.SKP, body_size)

    def visit_repeat_until_loop(self, node: RepeatUntilLoop) -> None:
        """repeat BODY until CONDITION: loop while condition is false"""
        body_start = len(self.instructions)
        self.visit(node.body)

        # Evaluate condition; last_cmp is set by the COM/AOP it generates
        self.visit(node.condition)

        # SKPT 1: skip the JMP if condition is TRUE (exit loop)
        self.emit(OpCode.SKPT, 1)
        # JMP back to body_start if condition is FALSE (keep looping)
        self.emit(OpCode.JMP, body_start)

    def visit_print_statement(self, node: PrintStatement) -> None:
        value = self.visit(node.expr)
        self.emit(OpCode.PRT, value)

    def visit_return_statement(self, node: ReturnStatement) -> None:
        if node.value:
            value = self.visit(node.value)
            self.emit(OpCode.RET, value)
        else:
            self.emit(OpCode.RET)

    def visit_function_statement(self, node: FunctionStatement) -> None:
        # Called when a FunctionStatement appears inside another body (unusual).
        # Just compile the inner body inline.
        self.visit(node.body)
        self.emit(OpCode.RET)

File: go/generator.go
package main

import (
	"fmt"
	"strconv"
)

type FuncEntry struct {
	StartIdx   int
	ParamNames []string
}

type Generator struct {
	instrs      []Instruction
	tempN       int
	funcTable   map[string]FuncEntry
	entryParams []string
}

func NewGenerator() *Generator {
	return &Generator{funcTable: make(map[string]FuncEntry)}
}

func (g *Generator) newTemp() string {
	s := fmt.Sprintf("t%d", g.tempN)
	g.tempN++
	return s
}

func (g *Generator) emit(op OpCode, operands ...string) {
	g.instrs = append(g.instrs, Instruction{op, operands})
}

func (g *Generator) Generate(ast *BlockNode) ([]Instruction, map[string]FuncEntry, []string, error) {
	g.instrs = nil
	g.tempN = 0
	g.funcTable = make(map[string]FuncEntry)
	g.entryParams = nil

	var funcs []*AlgoNode
	var bare []Node
	for _, s := range ast.Stmts {
		if a, ok := s.(*AlgoNode); ok {
			funcs = append(funcs, a)
		} else {
			bare = append(bare, s)
		}
	}

	if len(funcs) > 0 {
		entry := funcs[0]
		g.entryParams = entry.Params
		g.funcTable[entry.Name] = FuncEntry{len(g.instrs), entry.Params}
		if err := g.visitBlock(entry.Body); err != nil {
			return nil, nil, nil, err
		}
		g.emit(RET)
	}

	for _, s := range bare {
		if _, err := g.visit(s); err != nil {
			return nil, nil, nil, err
		}
	}

	return g.instrs, g.funcTable, g.entryParams, nil
}

func (g *Generator) visitBlock(b *BlockNode) error {
	for _, s := range b.Stmts {
		if _, err := g.visit(s); err != nil {
			return err
		}
	}
	return nil
}

func (g *Generator) visit(n Node) (string, error) {
	switch node := n.(type) {
	case *LiteralNode:
		return node.Value, nil
	case *IdentNode:
		return node.Name, nil
	case *ArrayLitNode:
		return g.visitArrayLit(node)
	case *ArrayAccessNode:
		return g.visitArrayAccess(node)
	case *BinaryOpNode:
		return g.visitBinaryOp(node)
	case *UnaryOpNode:
		return g.visitUnaryOp(node)
	case *AssignNode:
		return "", g.visitAssign(node)
	case *IfNode:
		return "", g.visitIf(node)
	case *WhileNode:
		return "", g.visitWhile(node)
	case *ForNode:
		return "", g.visitFor(node)
	case *RepeatNode:
		return "", g.visitRepeat(node)
	case *ReturnNode:
		return "", g.visitReturn(node)
	case *BlockNode:
		return "", g.visitBlock(node)
	case *AlgoNode:
		if err := g.visitBlock(node.Body); err != nil {
			return "", err
		}
		g.emit(RET)
		return "", nil
	}
	return "", fmt.Errorf("unknown node type %T", n)
}

func (g *Generator) visitArrayLit(node *ArrayLitNode) (string, error) {
	t := g.newTemp()
	elems := make([]string, len(node.Elems))
	for i, e := range node.Elems {
		v, err := g.visit(e)
		if err != nil {
			return "", err
		}
		elems[i] = v
	}
	ops := append([]string{t}, elems...)
	g.emit(ASN, ops...)
	return t, nil
}

func (g *Generator) visitArrayAccess(node *ArrayAccessNode) (string, error) {
	var arrName string
	if id, ok := node.Array.(*IdentNode); ok {
		arrName = id.Name
	} else {
		v, err := g.visit(node.Array)
		if err != nil {
			return "", err
		}
		arrName = v
	}
	idx, err := g.visit(node.Index)
	if err != nil {
		return "", err
	}
	t := g.newTemp()
	g.emit(IDX, arrName, idx, t)
	return t, nil
}

func (g *Generator) visitBinaryOp(node *BinaryOpNode) (string, error) {
	left, err := g.visit(node.Left)
	if err != nil {
		return "", err
	}
	right, err := g.visit(node.Right)
	if err != nil {
		return "", err
	}
	t := g.newTemp()
	switch node.Op {
	case "+", "-", "*", "/", "mod":
		g.emit(AOP, node.Op, left, right, t)
	default:
		g.emit(COM, node.Op, left, right, t)
	}
	return t, nil
}

func (g *Generator) visitUnaryOp(node *UnaryOpNode) (string, error) {
	operand, err := g.visit(node.Operand)
	if err != nil {
		return "", err
	}
	t := g.newTemp()
	g.emit(AOP, node.Op, operand, t)
	return t, nil
}

func (g *Generator) visitAssign(node *AssignNode) error {
	val, err := g.visit(node.Value)
	if err != nil {
		return err
	}
	switch target := node.Target.(type) {
	case *IdentNode:
		g.emit(ASN, target.Name, val)
	case *ArrayAccessNode:
		var arrName string
		if id, ok := target.Array.(*IdentNode); ok {
			arrName = id.Name
		} else {
			v, err := g.visit(target.Array)
			if err != nil {
				return err
			}
			arrName = v
		}
		idx, err := g.visit(target.Index)
		if err != nil {
			return err
		}
		g.emit(ASN, arrName+"["+idx+"]", val)
	default:
		return fmt.Errorf("invalid assignment target %T", node.Target)
	}
	return nil
}

func (g *Generator) visitIf(node *IfNode) error {
	if _, err := g.visit(node.Cond); err != nil {
		return err
	}
	skpIdx := len(g.instrs)
	g.emit(SKP, "0")
	thenStart := len(g.instrs)

	if err := g.visitBlock(node.Then); err != nil {
		return err
	}

	if node.Else != nil {
		jmpIdx := len(g.instrs)
		g.emit(JMP, "0")
		thenSize := jmpIdx - thenStart + 1
		g.instrs[skpIdx] = Instruction{SKP, []string{strconv.Itoa(thenSize)}}

		if err := g.visitBlock(node.Else); err != nil {
			return err
		}
		endIdx := len(g.instrs)
		g.instrs[jmpIdx] = Instruction{JMP, []string{strconv.Itoa(endIdx)}}
	} else {
		thenSize := len(g.instrs) - thenStart
		g.instrs[skpIdx] = Instruction{SKP, []string{strconv.Itoa(thenSize)}}
	}
	return nil
}

func (g *Generator) visitWhile(node *WhileNode) error {
	loopStart := len(g.instrs)
	if _, err := g.visit(node.Cond); err != nil {
		return err
	}
	skpIdx := len(g.instrs)
	g.emit(SKP, "0")
	bodyStart := len(g.instrs)

	if err := g.visitBlock(node.Body); err != nil {
		return err
	}
	g.emit(JMP, strconv.Itoa(loopStart))

	bodySize := len(g.instrs) - bodyStart
	g.instrs[skpIdx] = Instruction{SKP, []string{strconv.Itoa(bodySize)}}
	return nil
}

func (g *Generator) visitFor(node *ForNode) error {
	if err := g.visitAssign(node.Init.(*AssignNode)); err != nil {
		return err
	}
	loopVar := node.Init.(*AssignNode).Target.(*IdentNode).Name

	loopStart := len(g.instrs)
	endVal, err := g.visit(node.End)
	if err != nil {
		return err
	}
	tmpCmp := g.newTemp()
	g.emit(COM, "<=", loopVar, endVal, tmpCmp)

	skpIdx := len(g.instrs)
	g.emit(SKP, "0")
	bodyStart := len(g.instrs)

	if err := g.visitBlock(node.Body); err != nil {
		return err
	}

	tmpInc := g.newTemp()
	g.emit(AOP, "+", loopVar, "1", tmpInc)
	g.emit(ASN, loopVar, tmpInc)
	g.emit(JMP, strconv.Itoa(loopStart))

	bodySize := len(g.instrs) - bodyStart
	g.instrs[skpIdx] = Instruction{SKP, []string{strconv.Itoa(bodySize)}}
	return nil
}

func (g *Generator) visitRepeat(node *RepeatNode) error {
	bodyStart := len(g.instrs)
	if err := g.visitBlock(node.Body); err != nil {
		return err
	}
	if _, err := g.visit(node.Cond); err != nil {
		return err
	}
	g.emit(SKPT, "1")
	g.emit(JMP, strconv.Itoa(bodyStart))
	return nil
}

func (g *Generator) visitReturn(node *ReturnNode) error {
	if node.Value != nil {
		val, err := g.visit(node.Value)
		if err != nil {
			return err
		}
		g.emit(RET, val)
	} else {
		g.emit(RET)
	}
	return nil
}
